### PR TITLE
Use non-constant dx/dz

### PR DESCRIPTION
**Pull request by dschwoerer**
Potential solution for #8

### CODE REVIEW

**Comment by dschwoerer** [comment]
I have some MMS test for this - that are passing.
However, they require BOUT++ (python interface) - so not sure what is the best way / best place to add them ...

**Comment by bshanahan** [comment]
> However, they require BOUT++ (python interface) - so not sure what is the best way / best place to add them ...

Is there any way to have the tests without the python interface?

**Comment by dschwoerer** [comment]
It is possible, and I think I already have most of the code to do it in pure python. However that would test the python version, not BOUT++. Of course it could be done using the C++ interface as well, but building the python interface should be easy :-)

**Comment by dschwoerer** [comment]
I now added the test in the test branch.
As expected [it fails](https://github.com/boutproject/zoidberg/runs/6365014288):

case | convergence rate | l2-errors
---|---|---
const | 1.9979141144314099 | [0.25694862310799443, 0.07048700881896615, 0.018034505023958154, 0.004534781044690835, 0.0011353355724345777]
var1 | -0.0199579289182009 | [0.6419698336719127, 0.41228869188487777, 0.4563376663964457, 0.4796310743974576, 0.48631229290794287]
var2 | -0.0635603092360425 | [0.04248788059899262, 0.06933382171851286, 0.22246157978101588, 0.2688569192845526, 0.2809666527111458]
var3 | -0.05068870527866676 | [0.5090236738827739, 0.2757507814905903, 0.3360256795668088, 0.38159982232507494, 0.39524554596448]
var4 | -0.016657219168427675 | [0.4017088677256012, 0.22114742612361318, 0.23172290869711049, 0.2411329391716774, 0.2439331715599547]

Merging in this branch into test (test-dxdz), there are still some issues. Also merging in the w7x branch (#16 ) fixes them - and the tests [pass](https://github.com/boutproject/zoidberg/actions/runs/2299326510)

It seems that for a constant grid-spacing - dx and dz do not need to be correct (although maybe it needs to be constant?), but for a spatially varying grid that seems to be required - although I would assume we can rescale dx and dz arbitrarily ...

File: zoidberg/poloidal_grid.py [hunk 1]
```diff
@@ -33,11 +33,7 @@
     warnings.warn("Couldn't import matplotlib, plotting not available.")
     plotting_available = False
 
-try:
-    from . import rzline
-except:
-    # Python 2
-    import rzline
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply zoidberg/poloidal_grid.py:40]: Absolutely not -- Python 2 is dead!

**Review by ZedThree — APPROVED** [review]
I would perhaps change the `assert`s to more explicit errors

File: zoidberg/poloidal_grid.py [hunk 2]
```diff
@@ -473,42 +469,48 @@ def metric(self):
...
+                [
+                    [[J[j, i] * J[k, i] for i in range(2)] for j in range(2)]
+                    for k in range(2)
+                ]
+            ),
+            axis=2,
+        )
 
+        assert np.all(g[0, 0] > 0)
+        assert np.all(g[1, 1] > 0)
```
Comment (dschwoerer) [inline zoidberg/poloidal_grid.py:501]: @ZedThree would that be helpful:
```suggestion
        assert np.all(g[0, 0] > 0), f"g[0, 0] is expected to be positive, but some values are not (minimum {np.min(g[0, 0])})"
        assert np.all(g[1, 1] > 0), f"g[1, 1] is expected to be positive, but some values are not (minimum {np.min(g[1, 1])})"
```

Comment (ZedThree) [reply zoidberg/poloidal_grid.py:501]: Perfect, thanks!

File: zoidberg/poloidal_grid.py [hunk 3]
```diff
@@ -473,42 +469,48 @@ def metric(self):
...
+        g = g.transpose(2, 3, 0, 1)
+        assert np.all(np.linalg.det(g) > 0)
```
Comment (dschwoerer) [inline zoidberg/poloidal_grid.py:503]:
```suggestion
        assert np.all(np.linalg.det(g) > 0), f"All determinants of g should be positive, but some are not (minimum {np.min(np.linalg.det(g))})"
```

**Pull request merged**